### ISSUE TITLE
Network mode

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#12]
We need to design the network mode for this new API. 

In the points hooks API, network mode was available to provide the possibility for administrators to configure network-wide points hooks. These network-wide hooks would then fire on every site on the network. This would reduce duplication if they want to configure common hooks across the whole network.

I have always planned to carry on this great tradition in the new API. It is possible that network-wide hooks should be reevaluated entirely, but that is beyond the scope of this project.

I've compiled a list of places where `WordPoints_Points_Hook::get_instances()` is called. This is the main place that network-mode comes into play. 
- `WordPoints_Points_Hook::delete_callback()` mode (only single hook needed though)
- `WordPoints_Points_Hook::form_callback()` ditto
- `WordPoints_Points_Hook::update_callback()` ditto
- `WordPoints_Points_Hooks::get_handler()` ditto
- `WordPoints_Points_Hook::next_hook_id_number()` mode (may not be needed)
- `WordPoints_Points_Hook::get_description()` all (only single hook needed though)
- `WordPoints_Points_Hook::get_points()` ditto
- `WordPoints_Points_Hook::child::hook()` all
- `WordPoints_Post_Type_Points_Hook_Base::should_do_auto_reversals_for_post_type()` all (not needed though)

Reviewing this list, it looks like every time that only hooks of a single mode (standard or network) are retrieved, only a single hook is needed. The new API can introduce a method to get a single instance, which will improve this. If hook IDs are unique regardless of network status, we could ignore network-mode there altogether. However, I don't think that is realistic, so that method would still need to consider network mode in some way.

Two of the times that both modes of hook are retrieved, only a single hook is needed. The only time that we really need all hooks of a given type of a given hook is when the hook is being fired. So the only time that we need to retrieve all of the hooks of a given type is when we need both standard and network hooks. So we could change `get_instances()` to always return both.

The only possible exception is if we continue to store the points hooks as options, we may need to get just hooks of one single mode when generating the ID number for a hook. However, that will be an intrinsic part of the API anyway, sense all of the hooks of one type are stored in a single option anyway.

**Comment by JDGrimes** [comment]
We can also make network-wide support optional for a target/storage method.

**Comment by JDGrimes** [comment]
In the old API, an instance was identified by it's id number and the hook type. This meant that each hook was identified by two attributes, instead of just one. This caused lots of issues. The reason that the API was designed this way was because all of the hooks of a one type were stored together. This was inherited from the widgets API. It is because of this that all instances are retrieved when only a single one is needed.

In the new API, we need to make it so that hooks are identified by only a single attribute. This should be an ID number. The ID numbers were only unique to a hook type in the old API, but in the new API they will need to be unique across all hooks of a given type. 

They don't need to be unique across targets though. Different targets can use different storage methods, so unique IDs across all targets wouldn't be possible.

If we introduce a post storage method, it will work with this nicely, because each post is identified by just its ID. However, we'll need to modify the options storage method. Instead of storing all of the settings for the instances of one hook type together, each instance's settings would be stored in a separate option. However, we still need to be able to retrieve all of the instances of all types, and also just the instances of a single type. So we'll need to have an option that stores a list of all instance IDs and what type each is.

**Comment by JDGrimes** [comment]
Still to-do here:
- [x] Network Points Types screen.
- [x] Testing that network mode really works.

**Comment by JDGrimes** [comment]
> We can also make network-wide support optional for a target/storage method.

Currently I have the network mode toggles on the `WordPoints_Hooks` class, so it is global to all reactors. But I think we need to consider whether network mode should actually be a per-reactor setting. That would allow us to make network-mode support optional. It would also allow us to toggle each reactor independently. I'm not sure that is really a useful feature, but it might help avoid issues when reactions from multiple reactors are being displayed on the same screen (though I guess in many circumstances they'd probably both be either on or off anyway).

We could also go a step further, and remove network-mode entirely. I'm not sure that that is the best path forward, but I think it is worth considering. We would have separate methods for saving as well as for retrieving standard and network reactions. Then when you wanted to handle network reactions, you'd have to be explicit. It seems like that would probably lead to lots of duplicated logic though, both within the methods and at the different places that they are used. But if network support is going to be optional, we'll have to check before attempting to use network mode everywhere anyway. (I guess we actually wouldn't have to when we are only dealing with a specific reactor and know whether it supports network mode, but it would definitely become a necessity for any generic code.)

**Comment by JDGrimes** [comment]
> We could also go a step further, and remove network-mode entirely. I'm not sure that that is the best path forward, but I think it is worth considering. We would have separate methods for saving as well as for retrieving standard and network reactions. Then when you wanted to handle network reactions, you'd have to be explicit.

The benefit of network-mode though, is that it lets us use the same code for both standard and network reactions. For example, we have only once admin screen that is used both for the network points hooks and regular points hooks. So I don't think we should remove network-mode completely.

> It seems like that would probably lead to lots of duplicated logic though, both within the methods and at the different places that they are used. But if network support is going to be optional, we'll have to check before attempting to use network mode everywhere anyway.

So basically, if we make network-mode support optional, we'll also probably need to replace the global network-mode with a per-reactor network mode.

---

I have been thinking that maybe we should allow for the network and standard hooks to be handled independently for each reactor. I see no reason why they should have to use the same storage method, so we could let reactors that support network-mode define a storage class for the standard reactions and another one for the network reactions.

This might simplify some logic, because each of the storage methods wouldn't have to be constantly checking `$network_wide`. On the other hand, it might just lead to duplication, though that could probably be avoided with a little work.

Then network mode could be toggled within the magic getter of the reactor class, by just returning the network storage object instead of the standard one.

Note that we would still need to be able to get both standard and network reactions when firing events. So we would need to continue to provide some way of doing that.

**Comment by JDGrimes** [comment]
At the moment the network mode togglers still exist on the main hooks object, so network mode is still global. However, it isn't currently taken into account by the reactors, so we really don't have network mode at all. I'm going to leave it like this for now, because I'm not sure how best to proceed. When we start working on integrating with the current points hooks code (like the shortcodes, etc.) I think the way forward will become more clear. We might to away with network mode entirely, or we might make it a per-reactor thing. We'll see.

**Comment by JDGrimes** [comment]
Like I said above, getting rid of network mode is going to cause us to have to write lots of duplicate code. The only way that we could avoid duplicate logic would be to abstract all kinds of things out into their own functions that pass around the reaction storage object instead of a reference to the reactor itself.

For our purposes, we can break all of the code down into two types:
1. Code that needs to reference standard and/or network mode explicitly.
2. Code that just references whatever the current mode is.

This makes it clear that we need to have three different properties, one for standard reactions, one for network reactions, and one that will references either one based on the current network mode.

**Comment by JDGrimes** [comment]
## Global Network Mode?

We need to decide whether to have a global network mode, or to make it per-reactor. If we wanted to be really fancy we could even do both, with some kind of fallback to the default when per-reactor network mode isn't set explicitly. But therein lies the issue: network mode can become a complex monster really fast. What I'd like to do is make it as straightforward as possible, and so I think that in order to do that it should be really simple. So we might be better off just going with global network mode instead of trying to make it possible to toggle it per reactor. (Of course, there's nothing to stop reactors from introducing per-reactor network mode if they specifically need it.) Then we could make it more complex in the future only if it is really needed.

Another option would be to just implement network mode in the points reactor specifically, not as part of the general API.

The main issue with making network-mode global is that we wouldn't be able to make network mode optional. Actually, I guess there would be nothing stopping us from making it optional, but it might cause problems when network mode is enabled and no network storage class exists. Instead of returning null,  we could just return the standard storage object, but that would likely lead to worse trouble.

The concerns with global network mode is that we'll end up inadvertently affecting other reactors' behavior when settings network mode for the sake of one particularly. On the other hand, there's the chance that we'll forget to set network mode for all reactors involved by some specific code.

We could ask whether we'll really have much use for interacting reactors anyway. One concern was that we might write some generic code that could handle any reactor or set of reactors. But I doubt that would be a problem.

## Context-specific Network Mode

I think it comes down to this. Sometimes we want to reference network reactions specifically. But only in a certain context. The concern with global network mode is that we'd have network mode overflowing the context we set it in. For example, if we just turned on network mode for the entire network hooks screen, we might well inadvertently affect code that wasn't specifically related (maybe for a script that needs to show on every screen in the admin, or something like that). But on the other hand, sometimes there is other code that does specifically need to be tied to that context, perhaps even for a related reactor. That is where per-reactor network mode could bite us.

The best way to handle that would be to make network mode not be global, but not be reactor-specific either. Instead, it should be context specific. The network points type screen is one context, while firing an action is another (that uses both network and standard. but you get the idea), etc.

The question is whether we should develop an API to help capsulize these contexts, or whether we should just leave it up to each context to handle it separately. For the sake of keeping things clean and DRY, we should probably at least explore the possibility of making some kind of API for this.

There are several options. We could pass context objects around, or we could have a single toggle like we have now, but with a context arg that would be a string slug for each context. The former is sort of like what I talked about before, and it requires passing an object around everywhere, which isn't always pretty or easy. The single toggle in the global scope means that we don't have to pass objects around everywhere—or does it? We'd still need the context slug to be passed around so that any generic code would know what context it was in.

### Context Stack

New idea: what about a context switcher, and a context stack? When we enter a particular context, we push it onto a stack, and pop it off when we exit. Generic code would just use whatever the current mode was, but code that runs in a particular context would be able to set it specifically. It is, perhaps, a bit more complex than I'd like it to be, but it is simple enough at that.

I guess we'd need to be able to set the network mode for a particular context, but also set the current context itself. In other words, we might have code, like the points types screen, that sets the context and the network mode, and then calls some actions. Within those actions we might end up calling some code that switches context. So any code that needs to run specifically within the context of the network screen should declare that. Or maybe not, because then it wouldn't be reusable. But if we allow to specify the current context as being other than the current one on the stack, it would make the API more robust against failure to switch back correctly. Although I guess that code which wasn't context specific would still be subject to issues.

How exactly does this differ from just switching the global network mode on and off in each particular context? Why do we need named context, how is it different than just having a network-mode stack (which is slightly better than our current simple toggle, because we wouldn't have to worry about saving the previous value and resetting it).

One benefit of named contexts (though not the stack specifically) is that we can name some global contexts that might be useful, instead of each piece of code having to run the same logic. For example, we could have a `default` context, which would automatically be set to network-wide when in the network admin screens. (We might want to even hard-code these so they couldn't be messed with, but I guess there could be legitimate reasons to change them, like for an edge-case that isn't covered by the default logic.)

But to be able to use that sort of named context, we'd have to not have a stack, or to separate the stack from the toggles, as mentioned above.

### Issues

An issue with limiting network mode to a particular context, whether with basic toggles or a stack, is that the we have to be sure to set things back when we exit the current context. That can become annoying when the context has many exit points (many possible `return` points within a function for example). I'm not sure if there is any way to work around that in the design of the context API though, and I guess it is a universal issue to all of these possibilities. Though it would be nice if we could somehow tell a context to end as soon as the function returned or something. But that isn't possible. We could always refactor such a function so that the toggle/stack mods were outside of it instead, though. (And if we passed the context around, we wouldn't have this problem either.)

## Auto Network Mode

An idea that I've been playing around with in my mind is automatically setting network mode to on when in the network admin, and off otherwise. This could be a hard-coded thing that couldn't be changed (no public network-mode toggle) or it could be in addition one of these other solutions.

The truth is that we really shouldn't be using standard hooks at all in the network admin, without switching to a specific site. Because, although the network admin screens are loaded in the context of the "main site", that is sort of an "under-the-hood" thing. I'm not sure that we'd want to award points on the network admin screens, for example. We'd be awarding them in the context of the main site. I'm actually not sure what we should do about that. Should we abort all hook routing in network context completely? But what about some reactors that may find it useful?

We don't load non network-active modules in the network admin, for example, because they'll end up showing points hooks in the network points types screen that won't really fire across all sites (because the module may only be active on a few of the sites).

So should we have network-specific logs not in the context of a particular site? That is beyond the scope of what we're trying to do here I think. And then would the network points hooks be logged per-site or in the network logs?

Back to the original question at hand, an issue that we'd have with setting network mode automatically would be with reactors that don't support network mode. We've already talked about that above, we could just have the property value return null. But we haven't decided for sure whether to make this optional.

We can't mess with standard reactions when we're in the network admin anyway, because, as noted above, the modules aren't loaded like they are when we normally handle reactions for the main site. This could definitely cause odd things to happen.

So standard reactions cannot be loaded in the network admin.

I was thinking earlier that it might be useful to switch to standard mode when trying to display something in the context of a specific site but on the network screen (like the reaction count or something), in which case we'd switch to that site and toggle network mode off. But that can be dangerous because the modules for that site won't be loaded. Counting the reactions might be OK, but other things, like inadvertently firing them wouldn't be.

That said, I think it makes since to make this a default, but not to make it impossible to switch back to normal mode while in network context (assuming you know what you're doing).

Note that this also means that one of our original fears is relieved: we don't need to worry about being context-specific in the network admin; there is no chance for a clash of conflicts because everything will be network wide. (We won't make it impossible to switch out of network mode mode, as we said above, but we don't need to make it easy either, since there are definitely pitfalls.)

The only question would be whether we need to be able to switch to network mode when not in the network admin. That is the issue we need to resolve next.

**Comment by JDGrimes** [comment]
## Not Network Active

Another thing to consider: what about when WordPoints isn't network active? Should we still offer network mode in that case?

I'm actually not sure what we currently do in the points hooks API. But I can see that it might be beneficial to use network mode with some reactors even when WordPoints isn't network active—with the understanding that the network-wide reactions would only be firing on all sites where WordPoints was active, not all sites in the network. I say we might want to allow for that, but that isn't currently our use-case.

Here is a list of where we set the points hooks network mode:

| Where | Conditions | Mode | Reset? |
| --- | --- | --- | --- |
| `wordpoints_ajax_points_hooks_order()` | network screen nonce present (and user has caps) | on | no |
| `wordpoints_ajax_save_points_hook()` | network screen nonce present (and user has caps) | on | no |
| `hooks-no-js-load.php` | `is_network_admin()` | on | no |
| `hooks-load.php` | `is_network_admin()` (and user has caps) | on | no |
| `WordPoints_Points_Un_Installer` `::before_update()` | updating from before 1.9.0 | off | yes |
| `WordPoints_Points_Un_Installer` `::update_network_to_1_4_0()` | installed network wide | on | yes |
| `WordPoints_Points_Un_Installer` `::update_network_to_1_9_0()` | installed network wide | on | yes |
| `WordPoints_Points_Hook` `::points_type()` | current mode mismatch with ID passed | - | yes |
| `WordPoints_How_To_Get_Points_Shortcode` `::generate()` | `is_wordpoints_network_active()` | on | yes |

**Comment by JDGrimes** [comment]
For the Ajax requests, ~~I think that Ajax requests from the network admin will be in the context of the network admin, but I'm not 100% sure. If they are, then auto-setting the network admin would work fine for those.~~ See [WP#22589](https://core.trac.wordpress.org/ticket/22589).

Auto-setting would also work on the hook screens, which are actually loaded only when WordPoints is network active. (Actually it doesn't do that, but that's the way it should work. See https://github.com/WordPoints/wordpoints/issues/362.)

So for the first four cases in the above table, auto-setting should work (if we add a workaround for the network admin Ajax requests). We don't reset the mode back in any of those cases, so they essentially already work like the auto-default would.

The next three cases are for during when we are updating, which shows that we may want to be able to override the auto-set network mode (if we do auto-setting). Although, we really don't have to—we could/should probably refactor that so that it references the network or non-network reactions specifically.

The `WordPoints_Points_Hook::points_type()` instance isn't really important, since we don't have all that mess in the new API.

Much more interesting is the how-to-get-points shortcode usage. It could definitely be refactored to just have the points hooks or storage class passed in, but that would break back-compat.

There are definitely going to be other cases like this where we need to just get all of the reactions that affect a site, both network and standard (and maybe something else in future?). But rather than littering network mode toggles all over the place, it seems like it would be cleaner (and perhaps more forward-compatible) if we introduce methods for accessing all reactions. We already do this with `get_all_reactions_to_event()`, we could just introduce a `get_all_reactions()` as well, and that would mean no network-mode toggling.

The only problem I see is that that would force us to always retrieve all of the reactions, instead of maybe allowing a filter to turn off including the network ones, for example. But I guess if we needed to, we could still do that, we'd just need to reference each of the reaction storage objects specifically, in essence duplicating what is in the `get_all_*()` method. I guess we could actually add such a filter to the main methods, though it might not be as useful for this... but it would allow us to add hooks from elsewhere, as hinted above. (And for those cases where we want to filter, maybe we could pass a context parameter in.)

**Comment by JDGrimes** [comment]
Back to what to do when WordPoints isn't network active, we should default to not automatically include the network hooks when returning from the `get_all_*()` methods. That assumes that we go that route, of course, but I think this is a good argument in its favor: it consolidates the logic for when to include network hooks into a single location for each reactor.

This solves something that was bugging me about the toggler/stack idea: there's not really any way of preventing network mode from being turned on when it really shouldn't be. We have to check whether WordPoints is network-active each time, as in the table above. And if we get it wrong, as with the points hooks screen, we run into trouble.

**Comment by JDGrimes** [comment]
> For our purposes, we can break all of the code down into two types:
> 1. Code that needs to reference standard and/or network mode explicitly.
> 2. Code that just references whatever the current mode is.

There is actually a third case: when we need to get both.

This means that we may well need to provide several different solutions, one for each of these cases:
1. Explicit code: the `standard_reactions` and `network_reactions` properties.
2. Generic code (mode agnostic): the `reactions` property and auto-set mode.
3. Code that needs to get both: the `get_all_*()` methods.

This is a bit complex, perhaps, but I think it's about as simple as we can make it. The only thing that I should note is that we may still need to use toggles occasionally, like when updating, but mostly they would be an "under-the-hood" feature.

**Comment by JDGrimes** [comment]
For the record, the only significant place that we retrieve network mode, aside from the places that we reset it from the table above, is when we're determining the nonce to use for the Ajax requests from the hooks screen. And that will probably not be needed once we decide how to handle the network-admin originating Ajax requests.

This basically means that our code never really has to know what the current network mode is. The only time when we might need to modify it is when we are updating, and need to do something that involves the hook API. And even there, I'm wondering if we should just set some sort of flag that tells us when we're doing an update, and then just check for that flag in our network mode getter. However, I think that it would be simpler and more forward-compatible to just include a global network-mode getter and setter (or stack), but with a warning that it should be considered for internal use only. If you're going to mess with network mode you really need to know what you're doing.

I guess maybe it would be better if we also change what we call this. Instead of "network mode", why not "network-only mode"? Because, whereas before it was just a toggle, now it actually makes it impossible to access the standard reactions.

**Comment by JDGrimes** [comment]
So here are our action items:
- [x] Make it possible to reference the standard reactions explicitly.
- [x] Default network mode to the value of `is_network_admin()`.
- [x] Don't allow access to standard reactions in network mode.
- [x] Add a `get_all_reactions()` method.

So:
- Network mode will remain global.
- Whether a particular reactor behaves in the network admin when WordPoints isn't network-active is up to it to control.

We'll also need to integrate into the various places that the points hooks uses network mode.

**Comment by JDGrimes** [comment]
While working on this, I came across https://github.com/WordPoints/wordpoints/issues/363. The gist of it is that we have this special case when we're deleting a points type, where we need to delete all reactions for that points type across all sites on the network (if WordPoints is network-active). This is a special case, as I said, but we need to handle it nonetheless.

This is basically going to require us to switch to each site. I was thinking that maybe there would be a way to work around it, but I guess there isn't really, because we have to think about caching. But that being the case, I guess this just falls under the "need to reference explicitly" heading.

**Comment by JDGrimes** [comment]
## Getter Only

I'm seriously considering removing the network toggle feature. It just isn't that pretty, and makes more sense as a stack. And it isn't something that should be messed with. We do have the updates to think about (though we really don't _have_ to modify network mode there) but I'm not sure it is worth the potential future pain to developers to let them mess with something that doesn't need to be changed in 99.9% of cases.

So we could do a getter only, no setter.

For the update code we could just be explicit about referencing either the standard or network hooks, depending. In fact, it really doesn't matter at all because we won't be refactoring that to use the new API anyway: it is for updating the points hooks to the latest version, and will run before we import to the new API.

The only reason that we'd want to be able to toggle during updates is to force network mode when updating the network, just so nobody could accidentally get it wrong and forget to reference the network reactions specifically. I think that is the safest thing to do though.

And we also might need to do this for the Ajax requests in network admin context.

So we will be including a setter, but I think we should prefix it with an underscore to make it absolutely clear that it is a scary method.

**Comment by JDGrimes** [comment]
> - Whether a particular reactor behaves in the network admin when WordPoints isn't network-active is up to it to control.

Just to clarify, what we're saying here is that it is up to the code around each reactor to handle the case when WordPoints is installed on the main site (and thus loaded in the network admin) but isn't network active. 

Also, we're leaving open the possibility that the "network-wide" reactions could be created even when WordPoints wasn't network-active. This is useful when a reactor wants to offer firing across sites even when it is activated per-site. I'm not 100% sure that we should actually allow this by default, but I don't think we should make it completely impossible either. Perhaps we can find a way to let each reactor toggle this, but make it of by default.

**Comment by JDGrimes** [comment]
The only thing left here is to auto-set the network mode for Ajax requests, and to set it during updates.

**Comment by JDGrimes** [comment]
I have done the Ajax mode auto-setting, and the network hooks screen is working.

There are a few more things relating to that screen, but I'll handle them in #49. As var as the updates go, I've moved that to a new ticket as well: #48. We'll do new tickets for any other issues, too.
